src/Symfony/Component/Security/Csrf/TokenStorage/CookieTokenStorage.php

```diff
@@ -0,0 +1,165 @@
+     */
+    private $secure;
+
+    /**
+     * @param ParameterBag $cookies
+     * @param boolean $secure
+     */
+    public function __construct(ParameterBag $cookies, $secure)
+    {
+        $this->transientTokens = [];
+        $this->cookies = $cookies;
+        $this->secure = (bool) $secure;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \Symfony\Component\Security\Csrf\TokenStorage\TokenStorageInterface::getToken()
+     */
+    public function getToken($tokenId)
+    {
+        $token = $this->resolveToken($tokenId);
+
+        if (!strlen($token)) {
+            throw new TokenNotFoundException;
+        }
+
+        return $token;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \Symfony\Component\Security\Csrf\TokenStorage\TokenStorageInterface::hasToken()
+     */
+    public function hasToken($tokenId)
+    {
+        return strlen($this->resolveToken($tokenId)) > 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \Symfony\Component\Security\Csrf\TokenStorage\TokenStorageInterface::setToken()
+     */
+    public function setToken($tokenId, $token)
+    {
+        if (!strlen($token)) {
+            throw new \InvalidArgumentException('Empty tokens are not allowed');
+        }
+
+        $this->updateToken($tokenId, $token);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see \Symfony\Component\Security\Csrf\TokenStorage\TokenStorageInterface::removeToken()
+     */
+    public function removeToken($tokenId)
+    {
+        $token = $this->resolveToken($tokenId);
+
+        $this->updateToken($tokenId, '');
+
+        return strlen($token) ? $token : null;
+    }
+
+    /**
+     * @return array
+     */
+    public function createCookies()
+    {
+        $cookies = [];
+
+        foreach ($this->transientTokens as $tokenId => $token) {
+            // FIXME empty tokens are handled by the http foundations cookie class
+            // and are recognized as a "delete" cookie
+            // the problem is the that the value of deleted cookies get set to
+            // the string "deleted" and not the empty string
+
+            // TODO http only?
+            $name = $this->generateCookieName($tokenId);
+            $cookies[] = new Cookie($name, $token, 0, null, null, $this->secure, true);
+        }
+
+        return $cookies;
+    }
+
+    /**
+     * @param string $tokenId
+     * @return string
+     */
+    protected function resolveToken($tokenId)
+    {
+        if (isset($this->transientTokens[$tokenId])) {
+            return $this->transientTokens[$tokenId];
+        }
+
+        $name = $this->generateCookieName($tokenId);
+
+        return $this->cookies->get($name, '');
+    }
+
+    /**
+     * @param string $tokenId
+     * @param string $token
+     * @return void
+     */
+    protected function updateToken($tokenId, $token)
+    {
+        $token = (string) $token;
+        $name = $this->generateCookieName($tokenId);
+
+        if ($token === $this->cookies->get($name, '')) {
```

Comment:
The empty string token *is* the non-existant token. Thats because we dont have an explicit "delete" cookie header. If you want to delete a cookie, you need to overwrite a previously existing one. Thats also why it is disallowed to pass an empty string token to `setToken` (and the empty token makes no sense from a security point of view anyway).